compiler/rustc_middle/src/ty/util.rs

```diff
@@ -1002,6 +1060,42 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for OpaqueTypeExpander<'tcx> {
     }
 }
 
+struct WeakAliasTypeExpander<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    depth: usize,
```

Comment:
I'd like to note that I'm still using a plain depth counter over a hash set of `DefId`s to detect overflows since that feels more memory efficient to me.